tests/test/unify.rs

```diff
@@ -78,6 +78,48 @@ fn forall_equality() {
             InEnvironment { environment: Env([]), goal: '!2_0: '!2_1 }, \
             InEnvironment { environment: Env([]), goal: '!2_1: '!2_0 }]"
         }
+
+        goal {
+            // Function pointers with different ABIs should not be equal.
+            extern "Rust" fn(): Eq<extern "C" fn()>
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            // Function pointers with identical ABIs should be equal.
+            extern "Rust" fn(): Eq<extern "Rust" fn()>
+        } yields {
+            "Unique; substitution [], lifetime constraints []"
+        }
+
+        goal {
+            // Function pointers with different safety should not be equal.
+            unsafe fn(): Eq<fn()>
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            // Function pointers with identical safety should be equal.
+            unsafe fn(): Eq<unsafe fn()>
+        } yields {
+            "Unique; substitution [], lifetime constraints []"
+        }
+
+        goal {
+            // Variadic function pointers should not be equal to non-variadic fn pointers.
+            variadic fn(): Eq<fn()>
```

Comment:
Side note: the actual Rust syntax is `fn(...)` or something like that, maybe we should use that instead?